### PR TITLE
Node: tighten URL query type

### PR DESCRIPTION
**Pull request by OliverJAsh**
See the test I added which was previously failing.

Note:
- I had to use a faux `unknown` type since the types don't currently support TS 3.0. I noticed types can be specified as augmentations for TS 3.0 support, but I don't think it's possible to override an object's property in this way.
- I had to add a TS version comment due to weird lint errors in TS 2.0.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/node/node-tests.ts
Comment (OliverJAsh) [inline types/node/node-tests.ts]: For some reason `$ExpectError` doesn't seem to be working here:

```
$ npm run lint node

> definitely-typed@0.0.3 lint /Users/oliverjash/Development/DefinitelyTyped
> dtslint types "node"

Error: Errors in typescript@next for external dependencies:
../node-tests.ts(513,15): error TS2322: Type 'string | { [key: string]: {} | null | undefined; } | null | undefined' is not assignable to type 'string | ParsedUrlQuery | null | undefined'.
  Type '{ [key: string]: {} | null | undefined; }' is not assignable to type 'string | ParsedUrlQuery | null | undefined'.
    Type '{ [key: string]: {} | null | undefined; }' is not assignable to type 'string'.

    at /Users/oliverjash/Development/DefinitelyTyped/node_modules/dtslint/bin/index.js:210:19
    at Generator.next (<anonymous>)
    at fulfilled (/Users/oliverjash/Development/DefinitelyTyped/node_modules/dtslint/bin/index.js:5:58)
    at <anonymous>
```

File: types/node/url.d.ts
Comment (OliverJAsh) [inline types/node/url.d.ts]: To summarise the problem and fix:

``` ts
type ParsedUrlQuery = { [key: string]: string }

// Before
{
    declare const object: { [key: string]: any };
    // No error! :-(
    const parsedUrlQuery: ParsedUrlQuery = object;
}

// After
{
    declare const object: { [key: string]: unknown };
    // Error! :-(
    const parsedUrlQuery: ParsedUrlQuery = object;
}
```

File: types/node/index.d.ts
Comment (OliverJAsh) [inline types/node/index.d.ts]: The build now errors with:

> Error: accepts depends on node but has a lower required TypeScript version.

What to do?

The reason I had to add this was because without it, `npm run lint node` resulted in lots of weird errors. 🤷‍♂️ E.g.:

```
Error: /Users/oliverjash/Development/DefinitelyTyped/types/node/node-tests.ts:510:15
ERROR: 510:15  expect  Compile error in typescript@2.0 but not in typescript@2.1.
Fix with a comment '// TypeScript Version: 2.1' just under the header.
Type 'string' is not assignable to type 'UrlObject[]'.
ERROR: 510:15  expect  Compile error in typescript@2.0 but not in typescript@2.1.
Fix with a comment '// TypeScript Version: 2.1' just under the header.
Cannot redeclare block-scoped variable 'urlObjectQuery'.
…
```

These seem like invalid errors, and I'm not sure how to workaround this.

File: types/node/querystring.d.ts
Comment (OliverJAsh) [inline types/node/querystring.d.ts]: Alternatively we could do `{ [key: string]: string | string[] | number | number[] | boolean | boolean[]; }`, since these are all the types `querystring.stringify` and `url.format` will implicitly convert to a `string`:

``` sh-session
$ node
> require('querystring').stringify({ string: 'string', number: 1, boolean: true, numbers: [1,2], booleans: [true, false], strings: ['string1', 'string2'] })
'string=string&number=1&boolean=true&numbers=1&numbers=2&booleans=true&booleans=false&strings=string1&strings=string2'
```

IIUC, other value types like objects will be omitted:

``` sh-session
$ node
> require('querystring').stringify({ object: { foo: 1 } })
'object='
```

Currently with `unknown` we allow these types to be passed in—after all it doesn't generate a runtime error, so it is "safe".

https://devdocs.io/node/querystring#querystring_querystring_stringify_obj_sep_eq_options

File: types/node/index.d.ts
Comment (SimonSchick) [reply types/node/index.d.ts]: We cannot change the min version here as too many packages rely on it.

**Review by SimonSchick — CHANGES_REQUESTED** [review]

File: types/node/index.d.ts
Comment (SimonSchick) [reply types/node/index.d.ts]: Unless you are willing to update all dependant packages.

File: types/node/querystring.d.ts
Comment (SimonSchick) [reply types/node/querystring.d.ts]: I agree here, the `PoorMansUnknown` is a huge hack (as suggested by the name itself), we should wait for node typings to move beyond TS 3.x instead.

**Review by SimonSchick — CHANGES_REQUESTED** [review]

File: types/node/querystring.d.ts
Comment (OliverJAsh) [reply types/node/querystring.d.ts]: By agreeing, are you saying to replace `PoorMansUnknown` with `string | string[] | number | number[] | boolean | boolean[]`?

FWIW, I don't think it's a huge hack—there are no downsides to using it, other than the lack of a name. Even TypeScript itself uses it when it can't infer a generic.

I think we should consider the desired behaviour and then use that to pick between the two options here.

File: types/node/index.d.ts
Comment (OliverJAsh) [reply types/node/index.d.ts]: Can you advise on the 2.0 lint errors?

Comment (Flarna) [reply types/node/index.d.ts]: TS 2.1 has introduced keyof and Lookup Types which is needed for e.g. `url.UrlObject['query']`. You could use
```
const urlObjectQuery: string | null | querystring.ParsedUrlQueryInput = {};
// $ExpectError
const urlQuery: string | null | querystring.ParsedUrlQuery = urlObjectQuery;
```

Comment (OliverJAsh) [reply types/node/index.d.ts]: Thanks @Flarna. It doesn't test quite the same thing, but it will do!

File: types/node/querystring.d.ts
Comment (Flarna) [reply types/node/querystring.d.ts]: Someone may a "query builder class" which works with `PoorMansUnknown` but not with `string | string[] | number | number[] | boolean | boolean[]`:
```
class MyQuery {
    val: string;
    constructor(a: number) {
         this.val = `number${a}`;
    }
}
```

**Comment by OliverJAsh** [comment]
Ping @SimonSchick or someone else who can approve this

**Comment by SimonSchick** [comment]
Sorry I'm too biased here, I won't approve this with the `PoorMansUnknown` in it, we should aim to push the 3.x typings instead.

Please ping someone else if you wish this to be approved, feel free to ignore my rejection.

**Comment by OliverJAsh** [comment]
Can we get a consensus on this from other Node maintainers?

Options:

1. Ignore this bug for now, push to TS 3 typings, and then make this change afterwards (when we'll be able to use `unknown` instead of `PoorMansUnknown`)
2. Start benefitting from this change today, and continue to push for TS 3 typings

My two cents: 2 is the best option, given that `PoorMansUnknown` isn't that much of a hack/workaround: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33997#discussion_r266882699.

**Comment by Flarna** [comment]
I think waiting for TS 3.0 as the minimum typescript version is waste of time. There were several attempts to raise this version but even raising it to 2.1 (`keyof`, `Partial`) was not possible.

Another option to avoid the `PoorMansUnknown` would be to keep the current implementation for <3.2 and use an `unknown` based for >=3.2. See #34016 where I added a real type guard for >=3.2 and a function return a boolean for <3.2.

But I have no strong opinion here and I'm also fine with the solution as as. Maybe `PoorMansUnknown` should be inlined to `ParsedUrlQueryInput` (not sure if the `| unknown` is needed) to avoid that it spreads to other locations.

**Comment by OliverJAsh** [comment]
Note we're also using the poor man's `unknown` trick elsewhere in the Node typings: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33195/files#diff-a2f9a5377787f7084c7f52b20c0108cfR682.

> keep the current implementation for <3.2 and use an `unknown` based for >=3.2.

Hmm, interesting. However I'm not sure this gains us anything? Either way we're going to have `PoorMansUnknown` in the typings. 🤔 

> Maybe `PoorMansUnknown` should be inlined to `ParsedUrlQueryInput`

I'm happy to do that! I was trying to make the code self documenting so it would be clear that this needs to be changed to `unknown` when the TS upgrade eventually happens—but I realise this means it could leak into other usages. I'll inline it and add a comment instead.

**Comment by Flarna** [comment]
> Note we're also using the `PoorMansUnknown` trick elsewhere

Yes, and I can remember that there were quite some discussions about it at the time you raised this PR :o)

> Hmm, interesting. However I'm not sure this gains us anything? Either way we're going to have `PoorMansUnknown` in the typings. 🤔

With current I mean that one before your PR.

**Review by Flarna — APPROVED** [review]

**Review by BrunoScheufler — APPROVED** [review]

**Comment by OliverJAsh** [comment]
@tsd-bot @sandersn This one seems to have got stuck?

**Comment by OliverJAsh** [comment]
@RyanCavanaugh @sandersn This one seems to have got stuck?

**Pull request merged**
